Project.toml

```diff
@@ -9,10 +9,12 @@ ColorTypes = "3da002f7-5984-5a60-b8a6-cbb66c0b333f"
 Colors = "5ae59095-9a9b-59fe-a467-6f913c188581"
 DataStructures = "864edb3b-99cc-5e75-8d2d-829cb0a9cfe8"
 Dates = "ade2ca70-3891-5945-98fb-dc099432e06a"
+JuMP = "4076af6c-e467-56ae-b986-b466b2749572"
```

Comment:
As mentioned, this as a weak dependency / Requires.jl would be better I think.

Reply:
You mean JuMP or both MOI and JuMP ?

Reply:
No, I just added the comment to JuMP. I would prefer having

* MOI as a dependency and some `src/MOI_wrapper.jl` containing everything that can be done with MOI - maybe better in `src/wrappers/MOI.jl`? or even `interfaces` instead of `wrappers`?
* JuMP as a weak dependency / extra with an extension , probably called something like `ext/ManoptJuMPExt.jl` defining/extending Manopt/JuMP parts in there.

Since my understanding is (correct me if I am wrong) `MOI` is a lightweight interface while JuMP is a (maybe not really lightweight) modelling package?

Reply:
Oh, indeed MOI has even more dependencies than JuMP. Hm.
My worry is breaking changes that have downstream effects.

So let menorahs it that way:

If you feel that even both as an Extension is a reasonable amount of work, that would be cool.
Otherwise we keep this as is here and that is for me also fine.

If at some point we feel they are nicer as an extension, we might move them in a later PR.

Reply:
> My worry is breaking changes that have downstream effects.

MOI and JuMP have reached v1 and we are checking for breakage at every release, e.g., @odow just noticed some dependents are now failing when he checked to make a new release of MOI: https://github.com/jump-dev/MathOptInterface.jl/pull/2217#issuecomment-1601953083

I'm not so familiar yet with extensions. If it is written as an extension but not with Requires.jl, it will only work with Julia v1.9 then ? So we need both Requires.jl and Julia v1.9 extensions ? I'm kind of waiting for the new LTS and see how things go to go all in on extensions ^^

Reply:
We already went all-in with extensions and made them compatible with Requires.jl still.
So if you are on pre-1.9 we use Requires to load them, after that we use extensions. So you can “steal” most ideas from our main file.

Since our CI also runs tests on 1.8 (maybe even 1.6 if I am not confusing that with Manifolds?) so we still also test that both work.

So besides load times of Manopt, dependencies for me would be an argument to move them to weak dependencies/Require.jl

Reply:
Extensions were a little annoying to set up but after some initial hurdles they are working fine. Requires.jl fallback allows it to work on older Julia versions. Manifolds.jl is IMO a really good example how to set it up, including Requires.jl fallback and Documenter.jl support. Extensions significantly lowered our load times, which was the biggest gain.

Reply:
MOI is actually a pretty heavy dependency -- not necessarily in terms of number of dependent packages, but in terms of total package size. It's one of the largest Julia packages by number of methods.

> My worry is breaking changes that have downstream effects.

There will be no breaking changes in MOI. We test pretty heavily to avoid that.

Reply:
But then we should really go for a weak dependency here. I had though MOI (as an interface) would be lightweight. But if not, then I would really prefer a weak/Requires.jl thing.

Reply:
> I had though MOI (as an interface) would be lightweight

It's not really an interface. It's a modeling package. In some sense, JuMP is lightweight syntactic sugar for MOI, not the other way around.

MOI includes an automatic differentiation engine, a test suite with thousands of tests (in `/src` so that other packages can call `MOI.runtests(Manopt.Optimizer)`, not in `/test` to test MOI itself) , the automated bridging system, ... the list goes on. It's more of a mono-repo rather than the Julia approach of lots of small packages.

Reply:
Thanks for the clarification :)